src/main/java/com/atlassian/db/replica/internal/state/ConnectionState.java

```diff
@@ -262,7 +262,16 @@ private Connection prepareReadConnection(RouteDecisionBuilder decisionBuilder) t
             decisionBuilder.cause(writeConnection.getFirstCause().build());
             return writeConnection.get(decisionBuilder);
         }
-        if (consistency.isConsistent(() -> readConnection.get(decisionBuilder))) {
+        boolean isConsistent;
+        try {
+            isConsistent = consistency.isConsistent(() -> readConnection.get(decisionBuilder));
+        } catch (Exception e) {
+            if (!compatibleWithPreviousVersion) {
+                closeConnection(readConnection, decisionBuilder);
```

Comment:
This is an interesting behaviour, it will cause to close and create replica connections more often.
For CBR it will occur more often, while for WAW? not much probably?
WDYT?

Reply:
> This is an interesting behaviour, it will cause to close and create replica connections more often.
> For CBR it will occur more often, while for WAW? not much probably?
> WDYT?

It should never happen in our product as all our internal ReplicaConsistency implementations already handle exceptions. It could leak the connection if:
- ReplicaConsistency implementation doesn't catch it.
- and the usage of DualConnection is working only for happy path scenarios (for example connection is not closed in the finally block)

Reply:
If I understood correctly:

> and the usage of DualConnection is working only for happy path scenarios (for example connection is not closed in the finally block)

This is not happening as well in our product?

Reply:
> This is not happening as well in our product?

It's happening, but not often.